### PR TITLE
[c++] `PlatformConfig` refactoring

### PR DESCRIPTION
**Pull request by nguyenv**
**Issue and/or context:**

#2520

**Changes:**

* Add `column-major` as valid column ordering value
* Add missing `dense_nd_array_dim_zstd_level` option
* `tiledb_schema_from_arrow_schema` needs to take in additional `std::string soma_type` argument to ensure `PlatformConfig` parameters `dataframe_dim_zstd_level`, `dense_nd_array_dim_zstd_level`, and `sparse_nd_array_dim_zstd_level` get applied correctly
* Use the default `PlatformConfig` constructor if one has not been provided, not `std::nullopt`. This is why the default levels were not getting set correctly
* Apply default filter zstd (at TileDB default level) to attributes
* Filter names coming in from R API need to be respected
* `PlatformConfig` should be representable in JSON format; options for `FilterList`s should take in JSON nested arrays:
```
{
    "attr_name": {
         "filters": ["XOR", {"name": name="GZIP", COMPRESSION_LEVEL=3]
    }
}
```
* Need to set `FilterList` for `attrs` and `dims`
* Break up `tiledb_schema_from_arrow_schema` into more readable, separate functions
* Test passing values to `PlatformConfig` in pytest, ensuring defaults get overridden

### CODE REVIEW

File: libtiledbsoma/src/utils/arrow_adapter.cc
Comment (johnkerl) [inline libtiledbsoma/src/utils/arrow_adapter.cc:356]: Sorry @nguyenv , no -- this is not at all what I meant earlier.

* What I thought you meant was that we could/should have a single `platform_config.dim_zstd_level` which would be used for dataframe, sparse nd array, and dense nd array
  * And my answer was: no, we need separate configs per datatype
* What I was advocating for (however inarticulate I may have succeeded in being, despite my best efforts) was:
  * that we _do_ need separate:
    * `platform_config.dataframe_dim_zstd_level`
    * `platform_config.sparse_nd_array_dim_zstd_level`
    * `platform_config.dense_nd_array_dim_zstd_level`
  * that our not having had a `platform_config.dense_nd_array_dim_zstd_level` was an oversight that needs to be remedied

Comment (nguyenv) [reply libtiledbsoma/src/utils/arrow_adapter.cc:356]: OK understood - I can change this.

**Comment by johnkerl** [comment]
@nguyenv  I took the liberty of merging from latest `main` at https://github.com/single-cell-data/TileDB-SOMA/pull/2541/commits/42ea7e86fa7c8c2e2c2d4a4597e2ffbbf8caffd5 in order to bring in #2540 for CI ... heads up for your git-pulling information :)

**Review by eddelbuettel — APPROVED** [review]
It's a large-ish PR and still a little short on tests (outside the Python side I can't really focus on).  I think this deserves a qualified thumbs-up but it may be best to wait for another review or two.

**Pull request merged**
